src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+
     @Override
     public void add(T value) {
+        Node<T> current = new Node<>(null, value, null);
+        if (tail == null) {
+            tail = current;
+            head = tail;
+        } else {
+            current.prev = tail;
+            current.prev.next = current;
+            tail = current;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndexExistence(index);
+        Node<T> current;
+        if (index == FIRST_ELEMENT_INDEX) {
+            head = new Node<>(null, value, head);
+            tail = head;
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            current = getNodeByIndex(index);
+            current = new Node<>(current.prev, value, current);
+            current.next.prev = current;
+            current.prev.next = current;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> current = getNodeByIndex(index);
+        T oldObject = current.item;
+        current.item = value;
+        return oldObject;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> current = getNodeByIndex(index);
+        if (current.next == null && current.prev == null) {
+            head = null;
+            tail = null;
+        } else if (current.next == null) {
+            current.prev.next = null;
+            tail = current.prev;
+        } else if (current.prev == null) {
+            current.next.prev = null;
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if ((object == current.item)
+                    || (object != null && object.equals(current.item))) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == EMPTY_LIST_VALUE;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private MyLinkedList.Node<T> next;
+        private MyLinkedList.Node<T> prev;
+
+        Node(MyLinkedList.Node<T> prev, T element, MyLinkedList.Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
```

Comment:
I can suggest you to remove prev & next parameters in constructor. By default, you assigned prev & next in new Node  as null. So you don't need always write null in constructors)

Reply:
I disagree with this point, because this realization of inner class Node is made like as standart for LinkedList. The Node constructor takes 3 parameters under the hood. Not specifying them violates the concept of Node